filesystem/amlogic_emmc_partitions.ksy

```diff
@@ -0,0 +1,77 @@
+meta:
+  id: amlogic_emmc_partitions
+  title: Amlogic proprietary eMMC partition table
+  license: CC0-1.0
+  ks-version: 0.9
+  bit-endian: le
+  encoding: UTF-8
+  endian: le
+
+doc: |
+  This is an unnamed and undocumented partition table format implemented by
+  the bootloader and kernel that Amlogic provides for their Linux SoCs (Meson
+  series at least, and probably others). They appear to use this rather than GPT,
+  the industry standard, because their BootROM loads and executes the next stage
+  loader from offset 512 (0x200) on the eMMC, which is exactly where the GPT
+  header would have to start. So instead of changing their BootROM, Amlogic
+  devised this partition table, which lives at an offset of 36MiB (0x240_0000)
+  on the eMMC and so doesn't conflict. This parser expects as input just the
+  partition table from that offset. The maximum number of partitions in a table
+  is 32, which corresponds to a maximum table size of 1304 bytes (0x518).
+  
+doc-ref:
+  - http://aml-code.amlogic.com/kernel/common.git/tree/include/linux/mmc/emmc_partitions.h?id=18a4a87072ababf76ea08c8539e939b5b8a440ef
+  - http://aml-code.amlogic.com/kernel/common.git/tree/drivers/amlogic/mmc/emmc_partitions.c?id=18a4a87072ababf76ea08c8539e939b5b8a440ef
+
+seq:
+  - id: magic
+    contents: ["MPT", 0]
+  - id: version
+    size: 12
+    type: strz
+  - id: num_partitions
+    -orig-id: part_num
+    type: s4
+    valid:
+      min: 1
+      max: 32
+  - id: checksum
+    type: u4
+    doc: |
+      Unsigned modulo 32 sum of all the bytes in the "partitions" element below,
+      interpreted as a sequence of 32-byte little-endian unsigned integers.
```

Comment:
I checked this statement by calculating the checksum in the .ksy based on the instructions like this:

```diff
$ git diff --no-index amlogic_emmc_partitions\@{original,modified}.ksy
diff --git 1/amlogic_emmc_partitions@original.ksy 2/amlogic_emmc_partitions@modified.ksy
index 6d7f63c..19fda28 100644
--- 1/amlogic_emmc_partitions@original.ksy
+++ 2/amlogic_emmc_partitions@modified.ksy
@@ -40,12 +40,44 @@ seq:
     doc: |
       Unsigned modulo 32 sum of all the bytes in the "partitions" element below,
       interpreted as a sequence of 32-byte little-endian unsigned integers.
+  - size: 0
+    if: ofs_partitions < 0
   - id: partitions
     type: partition
     repeat: expr
     repeat-expr: num_partitions
+  - size: 0
+    if: len_partitions < 0
+
+instances:
+  ofs_partitions:
+    value: _io.pos
+  len_partitions:
+    value: _io.pos - ofs_partitions
+  checksum_steps:
+    pos: ofs_partitions
+    size: len_partitions
+    type: checksum_steps_type
+  checksum_expected:
+    value: checksum_steps.steps.last.val

 types:
+  checksum_steps_type:
+    seq:
+      - id: steps
+        type: 'checksum_step(_index != 0 ? steps[_index - 1].val : 0)'
+        repeat: eos
+  checksum_step:
+    -webide-representation: '{cur}'
+    params:
+      - id: prev_val
+        type: u4
+    seq:
+      - id: cur
+        type: u4
+    instances:
+      val:
+        value: '((prev_val + cur) & 0xffff_ffff) >> 0'
   partition:
     seq:
       - id: name
```

It turned out that the checksum calculated this way does not match the stored checksum in the test file [`amlogic-s905y2.bin`](https://github.com/kaitai-io/kaitai_struct_samples/blob/c3829d5/filesystem/amlogic_emmc_partitions/amlogic-s905y2.bin) that you provided:

<pre><code>├─ magic = [77, 80, 84, 0]
├─ version = <b>01.00.00</b>
├─ numPartitions = <b>0x12</b> <i>= 18</i>
├─ checksum = <ins><b>0x4337B596</b></ins> <i>= 1127724438</i>
▷─ partitions
├─ ofsPartitions = <b>0x18</b> <i>= 24</i>
├─ lenPartitions = <b>0x2D0</b> <i>= 720</i>
├─ checksumExpected = <ins><b>0x48EAC954</b></ins> <i>= 1223346516</i>
▷─ checksumSteps</code></pre>

So I looked up the actual code used in the Amlogic's code to generate the checksum ([`emmc_partitions.c:624`](http://aml-code.amlogic.com/kernel/common.git/tree/drivers/amlogic/mmc/emmc_partitions.c?id=18a4a87072ab#n624)) and ran it on the test file [`amlogic-s905y2.bin`](https://github.com/kaitai-io/kaitai_struct_samples/blob/c3829d5/filesystem/amlogic_emmc_partitions/amlogic-s905y2.bin). Here's the console output - for the first 3 partitions, individual checksum addends are printed (see source files in [**amlogic_emmc_partitions-checksum.zip**](https://github.com/kaitai-io/kaitai_struct_formats/files/7925542/amlogic_emmc_partitions-checksum.zip)):

```console
$ gcc -Wall -std=c11 main.c

$ ./a.exe
magic: 'MPT'
part_num: 18
partition[0], initial checksum: 0x0
  [10]: 0x746F6F62
  [9]: 0x64616F6C
  [8]: 0x00007265
  [7]: 0x0
  [6]: 0x00400000
  [5]: 0x0
  [4]: 0x0
  [3]: 0x0
  [2]: 0x0
  [1]: 0x0
partition[1], initial checksum: 0xD9115133
  [10]: 0x746F6F62
  [9]: 0x64616F6C
  [8]: 0x00007265
  [7]: 0x0
  [6]: 0x00400000
  [5]: 0x0
  [4]: 0x0
  [3]: 0x0
  [2]: 0x0
  [1]: 0x0
partition[2], initial checksum: 0xB222A266
  [10]: 0x746F6F62
  [9]: 0x64616F6C
  [8]: 0x00007265
  [7]: 0x0
  [6]: 0x00400000
  [5]: 0x0
  [4]: 0x0
  [3]: 0x0
  [2]: 0x0
  [1]: 0x0
partition[3], initial checksum: 0x8B33F399
partition[4], initial checksum: 0x644544CC
partition[5], initial checksum: 0x3D5695FF
partition[6], initial checksum: 0x1667E732
partition[7], initial checksum: 0xEF793865
partition[8], initial checksum: 0xC88A8998
partition[9], initial checksum: 0xA19BDACB
partition[10], initial checksum: 0x7AAD2BFE
partition[11], initial checksum: 0x53BE7D31
partition[12], initial checksum: 0x2CCFCE64
partition[13], initial checksum: 0x5E11F97
partition[14], initial checksum: 0xDEF270CA
partition[15], initial checksum: 0xB803C1FD
partition[16], initial checksum: 0x91151330
partition[17], initial checksum: 0x6A266463
         checksum: 0x4337B596
checksum_expected: 0x4337B596
```

As you can see, the checksum calculated by the routine matches the stored checksum. At the same time, it's visible what happens here - the checksum routine loops over the first partition over and over again, repeating for the number of partitions. Therefore, the resulting checksum is identical to `((0xD9115133 * 18) & 0xffff_ffff) = 0x4337B596` (run `'0x{:X}'.format((0xD9115133 * 18) & 0xffff_ffff)` in Python) - `0xD9115133` is the checksum of `partition[0]`, and this is repeated 18 times because there are 18 partitions.

Since the official implementation behaves this way (and all partition tables in this format must follow the same approach, otherwise the reference impl will reject them as invalid), I think it should also be documented in the `.ksy` so that the format spec reflects reality.

Reply:
Also your description contained some flaws, which would have to be corrected anyway:

> Unsigned ~~modulo 32~~ sum of all the bytes in the "partitions" element below,
> interpreted as a sequence of 32-~~byte~~ little-endian unsigned integers.

1. _Modulo 32_ would mean `% 32` (or `& 0b0001_1111`, alternatively `& 0x1f`), which is not the case - it is actually `& 0xffff_ffff` or `% (1 << 32)`, which is equivalent to `% (2 ** 32)`.
2. 32-<ins>**bit**</ins> integers, not 32-~~byte~~.

Reply:
Wow, great catch! Thanks for digging into that! I've updated the documentation to match what you describe and fix the typos. Should I also try to add checksum validation to the format like in your patch? Some of the features you use there are new to me, but I can read up on them.